modules/promotion/src/PromotionOrderProcessor.php

```diff
@@ -43,6 +43,11 @@ public function process(OrderInterface $order) {
     $order_type = $this->orderTypeStorage->load($order->bundle());
     $promotions = $this->promotionStorage->loadValid($order_type, $order->getStore());
     foreach ($promotions as $promotion) {
+      // Coupons have a separate order processor.
+      if (!$promotion->get('coupons')->isEmpty()) {
```

Comment:
Shouldn't we instead ensure that $this->promotionStorage->loadValid() doesn't return promotions that have coupons? We can filter them out at the query level, skipping the loading of potentially dozens of entities.